### ISSUE TITLE
API? how to do a "rolling groupby" or groupby with overlapping groups?

### ISSUE THREAD

**Issue by jorisvandenbossche** [pandas-dev/pandas#26959]
Sparked by @betatim's question on twitter: https://twitter.com/betatim/status/1141321049918906368

Suppose you have the following data:

```
repo_id = np.random.choice(np.arange(1000), 10000)
index = pd.Timestamp("2019-01-01") + pd.to_timedelta(np.random.randint(0, 24*30*6, size=10000), unit='H')
df = pd.DataFrame({'repo_id': repo_id}, index=index).sort_index()
```

```
In [111]: df.head() 
Out[111]: 
                     repo_id
2019-01-01 01:00:00      162
2019-01-01 01:00:00      850
2019-01-01 01:00:00      414
2019-01-01 02:00:00      753
2019-01-01 02:00:00      125
```

Data at somewhat randomly points in time (here hourly, but precision doesn't matter much).

Assume now you want to calculate a rolling 30D (monthly) statistic for each day.  
For example the rolling number of unique values over a period of time (the elegant but somewhat verbose python code to do this: https://gist.github.com/betatim/c59039682d92fab89859358e8c585313)

A rolling operation does not exactly give what you want, because `df.rolling()` will calculate this 24H mean for each row, while there might be many rows for a single day, and you are only interested in 1 value for that day.
A groupby/resample operation can also not achieve this goal, as that could give monthly number of unique values, but it cannot shift this month period one day at a time, since the groups cannot be overlapping.

With rolling you can do something like `df.rolling("30D").nunique().resample('D').last()`. But this is 1) very slow, as the `rolling` in the first step is doing way too many calculations that you afterwards throw away with `resample().last()`, and 2) I am not sure the binning for timestamps within the same day happens fully as desired.

I found this an interesting use case, to think about if there are ways to better handle this in pandas, or how such an API could look like.

**Comment by WillAyd** [comment]
I might be missing the point as I was having a hard time running code samples as is but wouldn't this do it for you?

```python
df.groupby(pd.Grouper(freq='1D')).count().rolling('30D').sum()
```

If you get the number of daily transactions first before doing the rolling op should be much more efficient

**Comment by jorisvandenbossche** [comment]
I think something like that was suggested as well on twitter, but: that does not take the unique values over a 30D period, but only for a single day. 
You cannot sum unique daily values to get unique monthly values .. (it's not just the *number* of "transactions", but the number of *unique* transactions. If it was only the number, yes, then that would have been additive).

**Comment by WillAyd** [comment]
So would #26958 solve this then? I think this will always be two steps (i.e. perform the rolling calc then resample / reshape as desired) from an API perspective

**Comment by jorisvandenbossche** [comment]
> So would #26958 solve this then? 

No, you can already do a "nunique" operation at the moment, but you need to type the verbose `.apply(pd.Series.nunique, raw=False)` for that. So I just opened that issue to make the shortcut `rollin().nunique()` also work.

> I think this will always be two steps (i.e. perform the rolling calc then resample / reshape as desired) from an API perspective

With the current API, yes, you need two steps. But that's also why I opened this issue, because doing it in two steps is very non-performant. The possible solution I mentioned above like:

```
df.rolling("30D").nunique().resample('D').last()
```

is kind of working, but is *much* slower than the function written by @betatim in https://gist.github.com/betatim/c59039682d92fab89859358e8c585313 (which just uses a python for loop over the groups that get selected with `.loc` and combine the results in a dataframe). 
This is because the rolling step is doing way too many calculations that you afterwards throw away with `resample().last()`

**Comment by WillAyd** [comment]
> The possible solution I mentioned above like:
> 
> ```
> df.rolling("30D").nunique().resample('D').last()
> ```
> 
> is kind of working, but is _much_ slower than the function written by @betatim in

Just to be clear this doesn't actually work right now right? There is no `unique` for rolling so did you just alias this to your apply op above?

If so wouldn't the bottleneck there just be the apply op? If that is implemented I don't think performance would be a concern.

From an API perspective I can't imagine right now that rolling should do any kind of resampling or reduction on its own.

**Comment by jorisvandenbossche** [comment]
Another way to put it: it's a rolling operation, but instead of it row-by-row (calculating the statistic moving the window one row at a time), your window shifts with bigger jumps. 
Only in this case, the jumps are not a regular number of rows, but related to a datetime frequency.

**Comment by jorisvandenbossche** [comment]
Sorry, my code snippet was indeed already assuming the other issue being fixed. The actual working one:

```
df.rolling("30D")['repo_id'].apply(lambda x: len(pd.unique(x)), raw=True).resample('D').last()
```

sorry for the confusion! 
will update the above

**Comment by pilkibun** [comment]
Thanks for posting this @jorisvandenbossche, it was a really fun problem.

One caution to anyone trying it, there's an understandable error in the given solution, caused by pandas' unpythonic treatment of time range slicing:
```python
# python
ar=list(range(10))
len(ar[:5])
5

# pandas
ix=DatetimeIndex(map(str,range(2000,2010)))
len(Series(index=ix).loc[:"2005"] )
6

# python
ar=list(range(1,5))
len(ar) # python
4

# pandas
len(pd.date_range("2001-01-01","2001-01-05",freq="D"))
5
```

Because of this pandas oddity, If you look at the groups generated by the `isoformat` slicing In the gist solution, you'll see that they include midnight of the day following "the window" if an event occured at that time. The upshot is that some random subset of the counts is slightly wrong. No big deal, but frustrating when you're checking your code against the "reference" solution ;-).

FWIW, here's the version I used:
```python
def n_active_repos(ser, period=30):
    """Unique active repos over the period

    With period=30 this gives monthly active repos
    With period=1 you get daily active repos
    """
    import datetime

    start = datetime.datetime(2019, 1, 1) 
    now = ser.index[-1]+pd.Timedelta("1D")
    days_since_start = (now - start).days

    data = []

    for n in range(days_since_start):
        s = start + datetime.timedelta(days=n)
        e = start + datetime.timedelta(days=n+period)

        if e > now:
            break
        # our "monthly" data
        periodic = ser.loc[
            (s <= ser.index) & 
            (ser.index < e)].sort_values() 
        data.append(dict(timestamp=e, repos=len(set(periodic))))

    return pd.DataFrame.from_records(data).set_index("timestamp").repos
``` 

Funnily enough `Series.truncate`  is billed as the idiomatic way to avoid the boolean test above,
```
Series.truncate
Truncate a Series or DataFrame before and after some index value.

This is a useful shorthand for boolean indexing based on index
values above or below certain thresholds.
```
...but has exactly the same behavior as the slicing version. maybe its keywords could be extended to handle this case (`upto` and `from` maybe?), since I assume fixing datetime string slicing is not an option.

**Comment by WillAyd** [comment]
> Another way to put it: it's a rolling operation, but instead of it row-by-row (calculating the statistic moving the window one row at a time), your window shifts with bigger jumps.
> Only in this case, the jumps are not a regular number of rows, but related to a datetime frequency.

I still don't totally see how this should fit into the API design of rolling; seems like it should still be two separate ops to me to roll and group, and in that case I think your other PR regarding nunique to get good performance covers the need here.

If I'm mistaken though can you maybe provide a smaller reproducible example that highlights the expected values?

**Comment by pilkibun** [comment]
- `groupy` uses a labels to segment the data into distinct groups.
- `rolling` uses either fixed or a variable (for example "30D") window, but the stride is always one row
- `expanding` grows the window size row by with a stride of 0

But this problem requires both a variable window and a variable stride.

**Comment by pilkibun** [comment]
I think `rolling("30D")` may be broken:
<details>
<summary>example code</summary>

```code
import pandas as pd
import random
from itertools import chain
N=7
index=pd.DatetimeIndex(chain(*[([f"2001-01-{i:02d}",]*random.randint(0,6))  for i in range(1,N)]))
ser=pd.Series(range(len(index)),index=index)

def debug_rolling(ser, spec):
    acc=[]
    def foo(ar):
        acc.append((int(ar[0]), len(ar)))
        #print(ar)
        return 0
    ser.rolling(spec).apply(foo,raw=True)
    offset,winsize=zip(*acc)
    return pd.DataFrame(dict(offset=offset,winsize=winsize))
    
print(debug_rolling(ser, 3))
print(debug_rolling(ser, pd.offsets.Day(3)))
```
</details>

Here's `ser.rolling(3)`, doing what I expected
<details>
<summary>slices used by `ser.rolling(3)` </summary>

```
    offset  winsize
0        0        3
1        1        3
2        2        3
3        3        3
4        4        3
5        5        3
6        6        3
7        7        3
8        8        3
9        9        3
10      10        3
11      11        3
12      12        3
13      13        3
14      14        3
15      15        3
16      16        3
17      17        3
```
</details>

And here's `ser.rolling("3D")`, which I have no idea what it's doing.
<details>
<summary>slices used by `ser.rolling("3D")` </summary>

```
    offset  winsize
0        0        1
1        0        2
2        0        3
3        0        4
4        0        5
5        0        6
6        0        7
7        0        8
8        0        9
9        1        9
10       4        7
11       4        8
12       4        9
13       4       10
14       9        6
15       9        7
16       9        8
17       9        9
18       9       10
19       9       11
```
</details>

**Comment by pilkibun** [comment]
Rolling time windows were introduced in 786edc7ed737 and it seems this behavior was by design.:
```
When using ``.rolling()`` with an offset. The offset is a time-delta. Take a backwards-in-time 
looking window, and aggregate all of the values in that window (including the end-point, but 
not the start-point). This is the new value at that point in the result. These are variable sized 
windows in time-space for each point of the input. You will get a same sized result as the input.
```

I've thought about this a bit more, and I think the current API is not limited, it's wrong.
By my reasoning, when you ask for `rolling(3)`, you're asking for a window of 3 *rows* and the engine rolls one *row* at a time so by analogy when you ask for `rolling("3D")`, since you're asking for a 3 *day* window, the engine should roll 1 *day* at a time.  

As a thought experiment,  if your data consists of minutely events, you may want hourly, daily, weekly,monthly, yearly summaries. Does it really make sense to give the user a yearly summary for every data point? I don't think so. The user probably either wants a year-by-year and then use groupby, or they would want a rolling yearly summary based on a coarser, time-based frequency, and not row by row. 

So I think `rolling("30D")` should do what betatim, I, and maybe most users would expect,
which is to treat offset windows as dictating a natural step unit. The old behavior could be supported
with a new `stride` argument, which takes either an offset or an integer. 

Examples:
- `rolling(int)`, unchanged
- `rolling("30D", stride=1)`, a 30 day window, rolling one row at a time (current behavior of  `rolling("30D")`)
- `rolling("30D")`, a 30 day window, rolling one (D)ay at a time
- `rolling("3M")`,  a 3 month  window, rolling one (M)onth at a time
- `rolling("356D")`, a yearly window, rolling one (D)ay at a time
- `rolling("7D", stride="1M")`, a 7 day window, rolling one month at a time, basically a stat
for the first week of very month, (assuming the first event is on the 1st of the month).
- `rolling(1000, stride='1M')`, *up to* first 1000 data points of each month. Or perhaps Unsupported.

The last two cases are a little contrived, but theyshow how expressive this is. `stride` is also good for perf, since it skips irrelevant data points.  More importantly,  I think this better aligns the default behavior with user expectations.

This could go through a deprecation cycle by starting to issue a warning asking users to set `stride=1` to maintain current behavior.

**Comment by WillAyd** [comment]
> I've thought about this a bit more and I think the current API isn't just limited, it's wrong.
> By my reasoning, when you ask for `rolling(3)`, you're asking for a window of 3 _rows_ and the engine rolls one _row_ at a time so by analogy when you ask for `rolling("3D")`, since you're asking for a 3 _day_ window, the engine should roll 1 _day_ at a time.

Is there a precedent for this in other applications? Taking Postgres as an example by their definition 

*A window function performs a calculation across a set of table rows that are somehow related to the current row*

https://www.postgresql.org/docs/current/tutorial-window.html

So curious to see what everyone else is doing.

Going back to the OP I'm still not entirely clear as to what is driving the larger discussion here. The main problem there seemed to be performance which is totally unrelated to the API design of rolling. 

Why do we want to cram both the rolling and reshaping into one operation?  By definition now rolling will return something that matches the size of the calling object, so with your proposals I _think_ we'd either be breaking that assumption or essentially broadcasting the result of those rolling operations across a frequency window. In either case I don't see why we wouldn't want to leave the latter operation to a resample or a transformation

**Comment by pilkibun** [comment]
@willayd, the best "solution" pandas for joris's example offers does ~10000 unique operations,
and throws away the all but 6 of the results. That's not resampling or transformation, just waste. It's why the tiny example dataset takes a *few seconds* to run. And that means pandas currently can't idiomatically handle this for "real" datasets. The bad perf is not due to python overhead, as joris has already said, but useless computation. The rolling code is already segmented into indexer creation and then computing the result based on the indexer. We're talking about implementing a new kind of indexer, and then obviously exposing it via the API. It's a fairly seamless extension. As for the "definition" of rolling, I don't think it's right to treat it like gospel. If the user asks for something else, it's ok to give it to them. If they wan t6 data points, it's reasonable to give them 6 rows back. But you could always compute 6 values and put them in a same-shaped nan-filled series, if that's the problem.

I don't much care what the SQL rquivalent looks like because I'm as certain that both cases are solvable, as I am that both solutions are subtle, verbose and unreadable. The exact reason why so many prefer pandas over SQL when possible.

**Comment by pilkibun** [comment]
Another point on the use of rolling
```python
df.rolling("30D")['repo_id'].apply(lambda x: len(pd.unique(x)), raw=True).resample('D').last()
```

The `rolling` engine determined the window by taking `Timedelta` backwards from the row's index label. That means that if the last event on "2019-01-30 00:07" for example, the window will include all events after `2018-12-31 00:07`, which is probably not what you want. and this will fluctuate randomly depending on the exact time of the last event on each day.

To fix this, quantize all Timestamps to end of day:

```python
period=30
# you can use days=2 to shift results to match the index returned by the original solution
ser.index=ser.index.floor("D")+pd.Timedelta(days=1,nanoseconds=-1) 
ser.rolling(f"{period}D").apply(lambda x: len(pd.unique(x)), raw=True).resample('D').last()[period-1:]
```

betatim's original solution doesn't have this issue, since it calculates Timedelta's relative to "2019-01-01".

This is really a fantastic learning problem.

**Comment by pilkibun** [comment]
The need for a `stride` argument to `rolling` has surfaced before, https://github.com/pandas-dev/pandas/issues/15354

**Comment by xuancong84** [comment]
Yup, I also encounter the same missing functionality. My solution in #28302 also addresses this issue.

Currently, pd.Grouper ignores _loffset_, it uses _base_ for split boundary offset.

Ideally, there needs to be 3 parameters, loffset (label_offset, default=0, the starting boundary offset of every group label), gstart (group_start, default=loffset, the starting offset of every group's data) and gspan (group_span, default=1, the span of every group's data).

For example, if we want to split a time-series data from 9:00am every day to 9:00am the next day, while keeping the group label at 0am (so that the datetime object reduces to date, i.e., 2020-05-26 00:00:00 => 2020-05-26), however, within each group we want the past 2 days of data (i.e., from 9am two days before that day til 9am on that day) in addition to that day's data (i.e., from 9am on that day til 9am the next day), [in other words, there will be duplicate entries among adjacent groups and the total number of data rows will be tripled], then we should call pd.Grouper(freq='D', loffset=0, gstart=9.0/24-2, gspan=3). This will give rise to:
`[('2020-01-01T00:00:00', <data from 2019-12-30T09:00:00 to 2020-01-02T09:00:00>), ('2020-01-02T00:00:00', <data from 2019-12-31T09:00:00 to 2020-01-03T09:00:00>),
('2020-01-03T00:00:00', <data from 2020-01-01T09:00:00 to 2020-01-04T09:00:00>),
('2020-01-04T00:00:00', <data from 2020-01-02T09:00:00 to 2020-01-05T09:00:00>),
...]`

**Comment by mbrb** [comment]
> Yup, I also encounter the same missing functionality. My proposed solution in #28302 also addresses this issue.
> 
> Currently, pd.Grouper ignores _loffset_, it uses _base_ for split boundary offset.
> 
> Ideally, there needs to be 3 parameters, loffset (label_offset, default=0, the starting boundary offset of every group label), gstart (group_start, default=loffset, the starting offset of every group's data) and gspan (group_span, default=1, the span of every group's data).
> 
> For example, if we want to split a time-series data from 9:00am every day to 9:00am the next day, while keeping the group label at 0am (so that the datetime object reduces to date, i.e., 2020-05-26 00:00:00 => 2020-05-26), however, within each group we want the past 2 days of data (i.e., from 9am two days before that day til 9am on that day) in addition to that day's data (i.e., from 9am on that day til 9am the next day), [in other words, there will be duplicate entries among adjacent groups and the total number of data rows will be tripled], then we should call pd.Grouper(freq='D', loffset=0, gstart=9.0/24-2, gspan=3). This will give rise to:
> `[('2020-01-01T00:00:00', <data from 2019-12-30T09:00:00 to 2020-01-02T09:00:00>), ('2020-01-02T00:00:00', <data from 2019-12-31T09:00:00 to 2020-01-03T09:00:00>), ('2020-01-03T00:00:00', <data from 2020-01-01T09:00:00 to 2020-01-04T09:00:00>), ('2020-01-04T00:00:00', <data from 2020-01-02T09:00:00 to 2020-01-05T09:00:00>), ...]`

Thanks for sharing your insights. Any idea to get the workaround working for the following scenario: https://stackoverflow.com/questions/62351499/how-to-group-pandas-dataframe-by-date-with-overlapping-sliding-window
